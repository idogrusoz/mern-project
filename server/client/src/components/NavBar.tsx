import {
    AppBar,
    Avatar,
    createStyles,
    fade,
    IconButton,
    InputBase,
    makeStyles,
    Menu,
    MenuItem,
    Theme,
    Toolbar,
    Typography,
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";
import React, { useContext, useState } from "react";
import { AuthContext } from "./auth/AuthContext";
import { useHistory } from "react-router-dom";
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        grow: {
            flexGrow: 1,
        },
        search: {
            position: "relative",
            borderRadius: theme.shape.borderRadius,
            backgroundColor: fade(theme.palette.common.white, 0.15),
            "&:hover": {
                backgroundColor: fade(theme.palette.common.white, 0.25),
            },
            marginRight: theme.spacing(2),
            marginLeft: 0,
            width: "100%",
            [theme.breakpoints.up("sm")]: {
                marginLeft: theme.spacing(3),
                width: "auto",
            },
        },
        searchIcon: {
            padding: theme.spacing(0, 2),
            height: "100%",
            position: "absolute",
            pointerEvents: "none",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
        },
        inputRoot: {
            color: "inherit",
        },
        inputInput: {
            padding: theme.spacing(1, 1, 1, 0),
            paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
            transition: theme.transitions.create("width"),
            width: "100%",
            [theme.breakpoints.up("md")]: {
                width: "20ch",
            },
        },
    }),
);
const NavBar = () => {
    const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
    const { signOut } = useContext(AuthContext);
    const history = useHistory();
    const classes = useStyles();
    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };

    const goToProfile = () => {
        setAnchorEl(null);
        history.push("/");
    };

    const goToAdd = () => {
        setAnchorEl(null);
        history.push("/add-post");
    };

    const handleSignOut = async () => {
        await signOut();

        setAnchorEl(null);
    };

    return (
        <AppBar>
            <Toolbar style={{ display: "flex", justifyContent: "space-between" }}>
                <div style={{ width: 100, display: "flex", justifyContent: "space-between", alignItems: "center" }}>
                    <Menu
                        id="simple-menu"
                        anchorEl={anchorEl}
                        keepMounted
                        open={Boolean(anchorEl)}
                        onClose={handleClose}
                    >
                        <MenuItem onClick={goToProfile}>Profile</MenuItem>
                        <MenuItem onClick={goToAdd}>AddPost</MenuItem>
                        <MenuItem onClick={handleSignOut}>SignOut</MenuItem>
                    </Menu>
                    <Typography variant="h6" style={{ fontWeight: 900 }}>
                        KERPICH
                    </Typography>
                </div>
                <div className={classes.search}>
                    <div className={classes.searchIcon}>
                        <SearchIcon />
                    </div>
                    <InputBase
                        placeholder="Searchâ€¦"
                        classes={{
                            root: classes.inputRoot,
                            input: classes.inputInput,
                        }}
                        inputProps={{ "aria-label": "search" }}
                    />
                </div>
                <IconButton color="secondary" aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
                    <Avatar />
                </IconButton>
            </Toolbar>
        </AppBar>
    );
};

export default NavBar;
