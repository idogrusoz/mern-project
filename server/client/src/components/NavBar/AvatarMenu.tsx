import { Avatar, IconButton, Menu, MenuItem } from "@material-ui/core";
import React, { FunctionComponent, useContext } from "react";
import { AuthContext } from "../Auth/AuthContext";

export type AvatarMenuProps = {
    handleClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
    anchorEl: HTMLElement | null;
    handleClose: () => void;
    goToProfile: () => void;
    goToAdd: () => void;
    handleSignOut: () => void;
};

const AvatarMenu: FunctionComponent<AvatarMenuProps> = ({
    handleClick,
    anchorEl,
    handleClose,
    goToAdd,
    goToProfile,
    handleSignOut,
}) => {
    const { user } = useContext(AuthContext);
    return (
        <>
            <IconButton color="secondary" aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
                <Avatar src={user ? user.image : undefined} />
            </IconButton>
            <Menu id="simple-menu" anchorEl={anchorEl} keepMounted open={Boolean(anchorEl)} onClose={handleClose}>
                <MenuItem onClick={goToProfile}>Profile</MenuItem>
                <MenuItem onClick={goToAdd}>AddPost</MenuItem>
                <MenuItem onClick={handleSignOut}>SignOut</MenuItem>
            </Menu>
        </>
    );
};

export default AvatarMenu;
