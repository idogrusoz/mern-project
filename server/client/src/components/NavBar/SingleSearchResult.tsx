import { Avatar, Grid, makeStyles, Typography } from "@material-ui/core";
import React, { FunctionComponent } from "react";
import { SearchedUser } from "../../../../interfaces/user.interfaces";

const useStyles = makeStyles({
    result: {
        marginTop: 10,
        "&:hover": {
            backgroundColor: "#ffffff",
            cursor: "pointer",
        },
    },
});

type SingleSearchResultProps = {
    user: SearchedUser;
};

const SingleSearchResult: FunctionComponent<SingleSearchResultProps> = ({ user }) => {
    const classes = useStyles();
    return (
        <Grid container alignItems="center" style={{ marginTop: 10 }} className={classes.result}>
            <Grid item xs={3}>
                <Avatar src={user.image} onClick={() => console.log(user._id)} />
            </Grid>
            <Grid item xs={9}>
                <Typography onClick={() => console.log(user._id)}>{user.userName}</Typography>
            </Grid>
        </Grid>
    );
};

export default SingleSearchResult;
