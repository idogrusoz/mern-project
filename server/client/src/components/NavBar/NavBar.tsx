import {
    AppBar,
    createStyles,
    fade,
    InputBase,
    makeStyles,
    Paper,
    Theme,
    Toolbar,
    Typography,
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";
import React, { useContext, useEffect, useRef, useState } from "react";
import { AuthContext } from "../Auth/AuthContext";
import { useHistory } from "react-router-dom";
import { SearchedUser } from "../../../../interfaces/user.interfaces";
import api from "../../api";
import AvatarMenu from "./AvatarMenu";
import SingleSearchResult from "./SingleSearchResult";
import { ProfileContext } from "../Profile/ProfileContext";
const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        grow: {
            flexGrow: 1,
        },
        search: {
            position: "relative",
            borderRadius: theme.shape.borderRadius,
            backgroundColor: fade(theme.palette.common.white, 0.15),
            "&:hover": {
                backgroundColor: fade(theme.palette.common.white, 0.25),
            },
            marginRight: theme.spacing(2),
            marginLeft: 0,
            width: "100%",
            [theme.breakpoints.up("sm")]: {
                marginLeft: theme.spacing(3),
                width: "auto",
            },
        },
        searchIcon: {
            padding: theme.spacing(0, 2),
            height: "100%",
            position: "absolute",
            pointerEvents: "none",
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
        },
        inputRoot: {
            color: "inherit",
        },
        inputInput: {
            padding: theme.spacing(1, 1, 1, 0),
            paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
            transition: theme.transitions.create("width"),
            width: "100%",
            [theme.breakpoints.up("md")]: {
                width: "20ch",
            },
        },
        result: {
            "&:hover": {
                backgroundColor: "#ffffff",
                cursor: "pointer",
            },
        },
        resultContainer: { position: "absolute", width: "100%", backgroundColor: "rgba(255, 255, 255, 0.7)" },
    }),
);
const NavBar = () => {
    const [searchResults, setSearchResults] = useState<SearchedUser[]>([]);
    const [displayResults, setDisplayResults] = useState<boolean>(false);

    const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
    const { signOut, user } = useContext(AuthContext);
    const { setProfileOwner } = useContext(ProfileContext);
    const history = useHistory();
    const classes = useStyles();
    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        setAnchorEl(event.currentTarget);
    };

    const pickUser = (user: SearchedUser) => {
        setProfileOwner(user);
        setDisplayResults(false);
        history.push(`/username/${user.userName}`);
    };

    const handleClose = () => {
        setAnchorEl(null);
    };

    const goToProfile = () => {
        setAnchorEl(null);
        setProfileOwner(null);
        history.push("/");
    };

    const goToAdd = () => {
        setAnchorEl(null);
        history.push("/add-post");
    };

    const handleSignOut = async () => {
        await signOut();

        setAnchorEl(null);
    };

    const handleChange = async (e: { target: { value: string } }) => {
        try {
            const results = await api.get(`users/search/${e.target.value}`);
            setSearchResults(results.data.data);
            setDisplayResults(results.data.data.length > 0);
        } catch (error) {
            setSearchResults([]);
        }
    };
    const resultsContainer = useRef<HTMLElement | undefined>();
    const pageClick = (e: any) => {
        if (resultsContainer.current && !resultsContainer.current!.contains(e.target)) {
            setDisplayResults(false);
        }
    };
    useEffect(() => {
        window.addEventListener("mousedown", pageClick, false);
    }, []);

    return (
        <AppBar>
            <Toolbar style={{ display: "flex", justifyContent: "space-around" }}>
                <Typography variant="h6" style={{ fontWeight: 900 }}>
                    KERPICH
                </Typography>
                {user && (
                    <div className={classes.search}>
                        <div className={classes.searchIcon}>
                            <SearchIcon />
                        </div>
                        <InputBase
                            placeholder="Searchâ€¦"
                            classes={{
                                root: classes.inputRoot,
                                input: classes.inputInput,
                            }}
                            inputProps={{ "aria-label": "search" }}
                            onChange={handleChange}
                        />
                        {displayResults && (
                            <Paper className={classes.resultContainer} ref={resultsContainer}>
                                {searchResults.map((user: SearchedUser, i: number) => {
                                    return <SingleSearchResult user={user} pickUser={pickUser} key={i} />;
                                })}
                            </Paper>
                        )}
                    </div>
                )}
                {user && (
                    <AvatarMenu
                        handleClick={handleClick}
                        handleClose={handleClose}
                        anchorEl={anchorEl}
                        goToProfile={goToProfile}
                        goToAdd={goToAdd}
                        handleSignOut={handleSignOut}
                    />
                )}
            </Toolbar>
        </AppBar>
    );
};

export default NavBar;
