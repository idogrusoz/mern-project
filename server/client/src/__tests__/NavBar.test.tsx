import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import NavBar from "../components/NavBar";

describe("NavBar test", () => {
    const wrapper = shallow(<NavBar />);
    it("renders without crash", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
