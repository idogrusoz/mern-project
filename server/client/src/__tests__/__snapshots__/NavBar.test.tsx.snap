// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar test renders without crash 1`] = `
<WithStyles(ForwardRef(AppBar))>
  <WithStyles(ForwardRef(Toolbar))
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-around",
      }
    }
  >
    <WithStyles(ForwardRef(Typography))
      style={
        Object {
          "fontWeight": 900,
        }
      }
      variant="h6"
    >
      KERPICH
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
